linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala

```diff
@@ -0,0 +1,3241 @@
+    }
+
+    if (t.tpe == NothingType)
+      fb += wa.Unreachable
+
+    t.tpe
+  }
+
+  /** Generates a vtable- or itable-based dispatch.
+   *
+   *  Before this code gen, the stack must contain the receiver and the args of the target method.
+   *  In addition, the receiver must be available in the local `receiverLocalForDispatch`. The two
+   *  occurrences of the receiver must have the type for dispatch.
+   *
+   *  After this code gen, the stack contains the result. If the result type is `NothingType`,
+   *  `genTableDispatch` leaves the stack in an arbitrary state. It is up to the caller to insert an
+   *  `unreachable` instruction when appropriate.
+   */
+  def genTableDispatch(receiverClassInfo: WasmContext.ClassInfo,
+      methodName: MethodName, receiverLocalForDispatch: wanme.LocalID): Unit = {
+    // Generates an itable-based dispatch.
+    def genITableDispatch(): Unit = {
+      fb += wa.LocalGet(receiverLocalForDispatch)
+      fb += wa.StructGet(
+        genTypeID.forClass(ObjectClass),
+        genFieldID.objStruct.itables
+      )
+      fb += wa.I32Const(receiverClassInfo.itableIdx)
+      fb += wa.ArrayGet(genTypeID.itables)
+      fb += wa.RefCast(watpe.RefType(genTypeID.forITable(receiverClassInfo.name)))
+      fb += wa.StructGet(
+        genTypeID.forITable(receiverClassInfo.name),
+        genFieldID.forMethodTableEntry(methodName)
+      )
+      fb += wa.CallRef(ctx.tableFunctionType(methodName))
+    }
+
+    // Generates a vtable-based dispatch.
+    def genVTableDispatch(): Unit = {
+      val receiverClassName = receiverClassInfo.name
+
+      fb += wa.LocalGet(receiverLocalForDispatch)
+      fb += wa.StructGet(
+        genTypeID.forClass(receiverClassName),
+        genFieldID.objStruct.vtable
+      )
+      fb += wa.StructGet(
+        genTypeID.forVTable(receiverClassName),
+        genFieldID.forMethodTableEntry(methodName)
+      )
+      fb += wa.CallRef(ctx.tableFunctionType(methodName))
+    }
+
+    if (receiverClassInfo.isInterface)
+      genITableDispatch()
+    else
+      genVTableDispatch()
+  }
+
+  private def genApplyStatically(t: ApplyStatically): Type = {
+    t.receiver.tpe match {
+      case NothingType =>
+        genTree(t.receiver, NothingType)
+        // nothing else to do; this is unreachable
+        NothingType
+
+      case NullType =>
+        genTree(t.receiver, NullType)
+        fb += wa.Unreachable // trap
+        NothingType
+
+      case _ =>
+        val namespace = MemberNamespace.forNonStaticCall(t.flags)
+        val targetClassName = {
+          val classInfo = ctx.getClassInfo(t.className)
+          if (!classInfo.isInterface && namespace == MemberNamespace.Public)
+            classInfo.resolvedMethodInfos(t.method.name).ownerClass
+          else
+            t.className
+        }
+
+        BoxedClassToPrimType.get(targetClassName) match {
+          case None =>
+            genTree(t.receiver, ClassType(targetClassName))
```

Comment:
In general I've chosen to leave NPE traps attached to the receiver, in terms of position handling. The rationale is to avoid too much back-and-forth when stepping through the code. Even though the NPE belongs to the application, in practice, having it report on the position of the receiver is just as good.